### DISCUSSION TITLE
Mounting sub-applications under APIRouter

### DISCUSSION THREAD

**Discussion by chriswhite199** [fastapi/fastapi#8682] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter
from starlette.testclient import TestClient

app = FastAPI()
api_router = APIRouter(prefix="/api")


@api_router.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


app.include_router(api_router)

subapi = FastAPI()


@subapi.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


api_router.mount("/subapi", subapi)

client = TestClient(app)

assert client.get('/api/app').status_code == 200

# this next assert fails
assert client.get('/api/subapi/sub').status_code == 200
```

### Description

Is it possible to mount a sub-application under an APIRouter? APIRouter itself has a mount function and accepts similar arguments to mounting a sub-application on a FastAPI instance, but I can't get the routing to actually work (nor can i get the openapi docs or spec to come back from that I would assume are the correct URLs.

The docs for sub applications note that the sub-application will have it's root_path correctly set, and I've tried a few combinations of manually setting the root_path on the subapi instance, but to no avail.

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

3.9.7

**Comment by chriswhite199** [comment]
As a work around, I can mount the sub application onto the root app and use the routers prefix

```python
app.mount(api_router.prefix + "/subapi", subapi)
```

But is there a cleaner way to do this directly on the APIRouter instance?

**Comment by STeveShary** [comment]
> But is there a cleaner way to do this directly on the APIRouter instance?

@chriswhite199 There is not.  The `ApiRouter` does have a function `mount` that is available from the base starlette class, but using that will not have your route added.   This become clear when we look at the `include_router` function.  If the mounted object is not a type of `APIRoute`, it will not be added to the list of routes on the application. You can see here: https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L660

You can include routers inside of other routers and that will use the prefix.  I have modified your example to show how to do this with routers:
```python3
from fastapi import FastAPI, APIRouter
from starlette.testclient import TestClient

app = FastAPI()
api_router = APIRouter(prefix="/api")


@api_router.get("/app")
def read_main():
    return {"message": "Hello World from main app"}



other_api_router = APIRouter(prefix="/subapi")


@other_api_router.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


api_router.include_router(other_api_router)
app.include_router(api_router)


client = TestClient(app)

assert client.get('/api/app').status_code == 200

# this next assert fails
assert client.get('/api/subapi/sub').status_code == 200
```

**Comment by chriswhite199** [comment]
@STeveShary - Thanks for the response. I already have APIRouters included in other APIRouters to provide a hierarchy, but I was specifically looking for mounting sub-applications (to sort-of-reverse-proxy a number of other apps under a /api/proxy/ base path).

In which case would it not be prudent to override the mount method in APIRouter (overriding the parent class's Router::mount method) and log.warning("There be dragons..") or raise a NotImplemented / some better exception?

**Comment by Kludex** [comment, marked as answer]
- Created https://github.com/tiangolo/fastapi/issues/10180
